exec/java-exec/src/main/java/org/apache/drill/exec/store/base/filter/FilterPushDownListener.java

```diff
@@ -0,0 +1,140 @@
+   * {@code <col name> <relop> <const>}, qualifies for push down for
+   * this scan.
+   * <p>
+   * If so, return an equivalent RelOp with the value normalized to what
+   * the plugin needs. The returned value may be the same as the original
+   * one if the value is already normalized.
+   *
+   * @param groupScan the scan element. Use {@code scan.getGroupScan()}
+   * to get the group scan
+   * @param relOp the description of the relational operator expression
+   * @return a normalized RelOp if this relop can be transformed into a filter
+   * push-down, @{code null} if not and thus the relop should remain in
+   * the Drill plan
+   * @see {@link ConstantHolder#normalize(org.apache.drill.common.types.TypeProtos.MinorType)}
+   */
+  RelOp accept(GroupScan groupScan, RelOp relOp);
+
+  /**
+   * Transform a normalized DNF term into a new scan. Normalized form is:
+   * <br><code><pre>
+   * (a AND b AND (x OR y))</pre></code><br>
+   * In which each {@code OR} term represents a scan partition. It
+   * is up to the code here to determine if the scan partition can be handled,
+   * corresponds to a storage partition, or can be done as a separate
+   * scan (as for a JDBC or REST plugin, say.)
+   * <p>
+   * Each term is accompanied by the Calcite expression from which it was
+   * derived. The caller is responsible for determining which expressions,
+   * if any, to leave in the query by returning a list AND'ed (CNF) terms
+   * to leave in the query. Those terms can be the ones passed in, or
+   * new terms to handle special needs.
+   *
+   * @param groupScan the scan node
+   * @param andTerms a list of the CNF (AND) terms, in which each is given
+   * by the Calcite AND node and the derived RelOp expression.
+   * @param orTerm the DNF (OR) term, if any, that includes the Calcite
+   * node for that term and the set of OR terms. Only provided if the OR
+   * term represents a simple list of values (all OR clauses are on the
+   * same column). The OR term itself is AND'ed with the CNF terms.
+   * @return a pair of elements: a new scan (that represents the pushed filters),
+   * and the original or new expression to appear in the WHERE clause
+   * joined by AND with any non-candidate expressions. That is, if analysis
+   * determines that the plugin can't handle (or cannot completely handle)
+   * a term, return the Calcite node for that term back as part of the
+   * return value and it will be left in the query. Any Calcite nodes
+   * not returned are removed from the query and it is the scan's responsibility
+   * to handle them. Either the group scan or the list of Calcite nodes
+   * must be non-null. Or, return null if the filter condition can't be handled
+   * and the query should remain unchanged.
+   */
+  Pair<GroupScan, List<RexNode>> transform(GroupScan groupScan,
+      List<Pair<RexNode, RelOp>> andTerms, Pair<RexNode, DisjunctionFilterSpec> orTerm);
```

Comment:
I'm concerned about such an approach. We will limit possible set of functions which may be applied for pushdown by functions defined in `RelOp.Op`. For example, if we would have some source with partitions, we wouldn't be able to push filter with conditions like this: `id + 10 = 20`. For example, for parquet, can do filter pruning for such conditions.

Reply:
I like the idea of returning both `GroupScan` and `List<RexNode>`, so the rule will be able either to create partial filter above scan or just use the scan or do nothing if null was returned. But I think it would be better in the listener's implementation to either use only range and equals conditions to be able to parallelize scan if it is possible or prune scan segments if we would have some metadata.

Reply:
> We should be able to push the filter t2.b = 10 into the scan for t2. Do you know if Drill already handles this case? If not, we should add it.

Yes, I believe Drill able to do such optimizations (at least create the corresponding filter considering join condition), it was added in the scope of [DRILL-6173](https://issues.apache.org/jira/browse/DRILL-6173).

Reply:
Regarding addressing this issue in the scope of another Jira, I don't think that it is a good idea since there are some PRs that are waiting for this one, and they may extend this interface, so updating its signature later and changing the approach will force us to rewrite its implementations.

Reply:
> So, it seemed to me that, in general, different plugins might use different strategies, or combine strategies.
> 
> Is this what you addressed with your comment, or am I missing something?

Yes, I meant to provide plugin the ability to specify its own strategy instead of enforcing the existing one.

Ideally, I thought about implementing abstraction which will be able to cover the existing parquet filter pushdown strategy.

Reply:
@vvysotskyi Depending on the scope of [DRILL-7557](https://issues.apache.org/jira/projects/DRILL/issues/DRILL-7557), I think it's fine to get this initial enhancement committed so that we can start using it in storage plugins. My plan at least is to start with very simple rules to begin with.

Regardless, IMHO, this is a very important PR to get committed as it will greatly facilitate plugin development.

Reply:
@cgivre, I agree that this one is a very important change and will help to simplify developing new plugins.
My objection is when we will merge it as it is, the scope of DRILL-7557 will become larger since except addressing those issues, we would have to modify the implementations which would use a provided interface.

Reply:
@vvysotskyi and @cgivre - all good points. There is nothing special about the strategy offered; it is just a set of rules implemented using Calcite. Other plugins are free to use their own rules (or copy/paste these rules, or extend them.)

The point is valid: we should commit something we agree upon. So, I'm going to rethink this to figure out how to handle the Parquet case. Maybe we need to address the underlying issues first (missing algebra rewrites in Drill) to see can avoid making the API too wide.

My fear is, if we simply expose all of the current complexity, then there is nothing to abstract; we might as well continue to copy/paste big wads of code so each plugin can make the few tweaks that are needed.

Just a matter of finding the time to understand all the use cases well enough to see what might be in common. I'll offer a proposal in a few days.